File: native_host/notes_bridge.py
#!/usr/bin/env python3
import json
import mimetypes
import os
import re
import struct
import subprocess
import sys
import tempfile
import time
import uuid
from pathlib import Path
from urllib.parse import unquote, urlparse
import urllib.request
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple


HOST_NAME = "com.codex.apple_notes_webclipper"
CACHE_DIR = Path(tempfile.gettempdir()) / "apple-notes-webclipper"
CACHE_TTL_SECONDS = 7 * 24 * 60 * 60


def _ensure_cache_dir() -> None:
    CACHE_DIR.mkdir(parents=True, exist_ok=True)


def _cleanup_cache_dir() -> None:
    try:
        _ensure_cache_dir()
        now = int(time.time())
    except Exception:
        return

    try:
        for p in CACHE_DIR.iterdir():
            try:
                if not p.is_file():
                    continue
                age = now - int(p.stat().st_mtime)
                if age > CACHE_TTL_SECONDS:
                    p.unlink()
            except Exception:
                continue
    except Exception:
        return


def _read_exactly(n: int) -> bytes:
    buf = b""
    while len(buf) < n:
        chunk = sys.stdin.buffer.read(n - len(buf))
        if not chunk:
            raise EOFError()
        buf += chunk
    return buf


def read_native_message() -> Optional[Dict[str, Any]]:
    try:
        raw_len = _read_exactly(4)
    except EOFError:
        return None
    msg_len = struct.unpack("<I", raw_len)[0]
    if msg_len <= 0:
        return None
    raw = _read_exactly(msg_len)
    return json.loads(raw.decode("utf-8"))


def send_native_message(payload: Dict[str, Any]) -> None:
    encoded = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    sys.stdout.buffer.write(struct.pack("<I", len(encoded)))
    sys.stdout.buffer.write(encoded)
    sys.stdout.buffer.flush()


def run_osascript_file(script_path: str, argv: List[str], timeout_s: int = 60) -> str:
    proc = subprocess.run(
        ["osascript", script_path, *argv],
        check=False,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=timeout_s,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or f"osascript failed (code {proc.returncode})")
    return proc.stdout.strip()


def run_jxa(code: str, timeout_s: int = 60) -> str:
    proc = subprocess.run(
        ["osascript", "-l", "JavaScript", "-"],
        input=code,
        check=False,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=timeout_s,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or f"JXA failed (code {proc.returncode})")
    return proc.stdout.strip()


def escape_html(text: str) -> str:
    return (
        text.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&#39;")
    )


def markdown_with_data_images_to_notes_html(markdown: str) -> str:
    lines = markdown.splitlines()
    out: List[str] = []
    leading_ws_re = re.compile(r"^[ \t]+")
    for line in lines:
        if not line.strip():
            out.append("<div><br></div>")
            continue
        if line.lstrip().startswith("<img ") and line.rstrip().endswith(">"):
            out.append(f"<div>{line.strip()}</div>")
            continue
        m = leading_ws_re.match(line)
        if m:
            ws = m.group(0)
            rest = line[len(ws) :]
            ws_html = ws.replace("\t", "    ").replace(" ", "&nbsp;")
            out.append(f"<div>{ws_html}{escape_html(rest)}</div>")
        else:
            out.append(f"<div>{escape_html(line)}</div>")
    return "\n".join(out)


@dataclass
class ImageJob:
    token: str
    url: str
    alt: Optional[str] = None


def _guess_mime_from_url(url: str) -> str:
    lower = url.lower()
    if lower.endswith(".png"):
        return "image/png"
    if lower.endswith(".jpg") or lower.endswith(".jpeg"):
        return "image/jpeg"
    if lower.endswith(".gif"):
        return "image/gif"
    if lower.endswith(".webp"):
        return "image/webp"
    return "application/octet-stream"

def _ext_from_mime_or_url(mime: str, url: str) -> str:
    mime = (mime or "").lower()
    mapping = {
        "image/png": ".png",
        "image/jpeg": ".jpg",
        "image/jpg": ".jpg",
        "image/gif": ".gif",
        "image/webp": ".webp",
    }
    if mime in mapping:
        return mapping[mime]
    guessed = mimetypes.guess_extension(mime) if mime else None
    if guessed:
        return guessed
    try:
        suffix = Path(urlparse(url).path).suffix
        if suffix and len(suffix) <= 6:
            return suffix
    except Exception:
        pass
    return ".bin"


def _read_file_url_bytes(file_url: str) -> Tuple[bytes, str]:
    u = urlparse(file_url)
    path = unquote(u.path)
    data = Path(path).read_bytes()
    mime = _guess_mime_from_url(path)
    return data, mime


def cache_image_to_file_url(url: str, referer: Optional[str]) -> str:
    _ensure_cache_dir()
    if url.startswith("file://"):
        data, mime = _read_file_url_bytes(url)
    else:
        data, mime = download_bytes(url, referer=referer)

    ext = _ext_from_mime_or_url(mime, url)
    filename = f"img-{uuid.uuid4().hex}{ext}"
    path = CACHE_DIR / filename
    path.write_bytes(data)
    return path.as_uri()


def download_bytes(url: str, referer: Optional[str]) -> Tuple[bytes, str]:
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X) AppleWebKit/537.36 (KHTML, like Gecko) Chrome Safari",
            **({"Referer": referer} if referer else {}),
        },
        method="GET",
    )
    with urllib.request.urlopen(req, timeout=60) as resp:
        data = resp.read()
        content_type = resp.headers.get("Content-Type") or ""
        mime = content_type.split(";")[0].strip() if content_type else ""
        if not mime:
            mime = _guess_mime_from_url(url)
        return data, mime


def replace_image_tokens(markdown: str, images: List[ImageJob], source_url: Optional[str]) -> str:
    out = markdown
    for img in images:
        placeholder = f"[[[IMG:{img.token}]]]"
        try:
            file_url = cache_image_to_file_url(img.url, referer=source_url)
            replacement = f'<img src="{file_url}">'
        except Exception:
            replacement = img.url
        out = out.replace(placeholder, replacement)
    return out


def render_html_preview(payload: Dict[str, Any]) -> Dict[str, Any]:
    title = str(payload.get("title") or "Untitled")
    source_url = str(payload.get("sourceUrl") or "")
    markdown = str(payload.get("markdown") or "").strip()

    images_raw = payload.get("images") or []
    images: List[ImageJob] = []
    for item in images_raw:
        try:
            images.append(ImageJob(token=str(item.get("token")), url=str(item.get("url")), alt=item.get("alt")))
        except Exception:
            continue

    if images:
        markdown = replace_image_tokens(markdown, images, source_url)

    fragment = markdown_with_data_images_to_notes_html(markdown)
    full = "\n".join(
        [
            "<!doctype html>",
            "<html>",
            "<head>",
            '  <meta charset="utf-8">',
            f"  <title>{escape_html(title)}</title>",
            "</head>",
            "<body>",
            fragment,
            "</body>",
            "</html>",
        ]
    )

    _ensure_cache_dir()
    out_path = CACHE_DIR / f"preview-{uuid.uuid4().hex}.html"
    out_path.write_text(full, encoding="utf-8")
    return {"ok": True, "htmlPath": str(out_path), "htmlFileUrl": out_path.as_uri()}


def list_folders() -> Dict[str, Any]:
    code = r"""
      const Notes = Application('Notes');
      Notes.includeStandardAdditions = true;

      function collect(folder, prefix) {
        const name = folder.name();
        const path = prefix ? `${prefix}/${name}` : name;
        let out = [{ path }];
        const children = folder.folders();
        for (const child of children) out = out.concat(collect(child, path));
        return out;
      }

      const accounts = Notes.accounts().map(acc => {
        const folders = [];
        for (const f of acc.folders()) folders.push(...collect(f, ''));
        // Remove special folders if present.
        const filtered = folders.filter(x => x.path !== 'Recently Deleted');
        return { name: acc.name(), folders: filtered };
      });

      JSON.stringify({ ok: true, accounts });
    """
    raw = run_jxa(code)
    return json.loads(raw)


def create_note(payload: Dict[str, Any], script_dir: str) -> Dict[str, Any]:
    title = str(payload.get("title") or "Untitled")
    source_url = str(payload.get("sourceUrl") or "")
    markdown = str(payload.get("markdown") or "").strip()
    folder = payload.get("folder") or {}
    account_name = str(folder.get("accountName") or "")
    folder_path = str(folder.get("folderPath") or "")

    images_raw = payload.get("images") or []
    images: List[ImageJob] = []
    for item in images_raw:
        try:
            images.append(ImageJob(token=str(item.get("token")), url=str(item.get("url")), alt=item.get("alt")))
        except Exception:
            continue

    if images:
        markdown = replace_image_tokens(markdown, images, source_url)

    html = markdown_with_data_images_to_notes_html(markdown)

    with tempfile.NamedTemporaryFile("w", delete=False, suffix=".html", encoding="utf-8") as f:
        f.write(html)
        html_path = f.name

    try:
        script_path = os.path.join(script_dir, "scripts", "create_note.applescript")
        note_id = run_osascript_file(script_path, [account_name, folder_path, title, html_path], timeout_s=120)
        return {"ok": True, "noteId": note_id}
    finally:
        try:
            os.unlink(html_path)
        except OSError:
            pass


def main() -> None:
    script_dir = os.path.dirname(os.path.abspath(__file__))
    _cleanup_cache_dir()

    while True:
        msg = read_native_message()
        if msg is None:
            return

        action = msg.get("action")
        try:
            if action == "ping":
                send_native_message({"ok": True})
            elif action == "listFolders":
                send_native_message(list_folders())
            elif action == "createNote":
                send_native_message(create_note(msg, script_dir))
            elif action == "renderHtml":
                send_native_message(render_html_preview(msg))
            else:
                send_native_message({"ok": False, "error": f"Unknown action: {action}"})
        except Exception as e:
            send_native_message({"ok": False, "error": str(e)})


if __name__ == "__main__":
    main()

File: native_host_macos/NotesBridge.swift
import Foundation

private let hostName = "com.codex.apple_notes_webclipper"
private let cacheDirName = "apple-notes-webclipper"
private let cacheTtlSeconds: TimeInterval = 7 * 24 * 60 * 60

private enum NativeHostError: Error, CustomStringConvertible {
  case invalidMessage(String)
  case io(String)
  case process(String)
  case timeout(String)
  case download(String)

  var description: String {
    switch self {
    case .invalidMessage(let msg): return msg
    case .io(let msg): return msg
    case .process(let msg): return msg
    case .timeout(let msg): return msg
    case .download(let msg): return msg
    }
  }
}

private func readExactly(_ byteCount: Int, from handle: FileHandle) throws -> Data? {
  var data = Data()
  while data.count < byteCount {
    let chunk = try handle.read(upToCount: byteCount - data.count) ?? Data()
    if chunk.isEmpty {
      if data.isEmpty { return nil }
      throw NativeHostError.io("Unexpected EOF")
    }
    data.append(chunk)
  }
  return data
}

private func readNativeMessage() throws -> [String: Any]? {
  guard let header = try readExactly(4, from: .standardInput) else { return nil }
  var lengthLE: UInt32 = 0
  _ = withUnsafeMutableBytes(of: &lengthLE) { header.copyBytes(to: $0) }
  let length = Int(UInt32(littleEndian: lengthLE))
  if length <= 0 { return nil }

  guard let body = try readExactly(length, from: .standardInput) else { return nil }
  let obj = try JSONSerialization.jsonObject(with: body, options: [])
  guard let dict = obj as? [String: Any] else {
    throw NativeHostError.invalidMessage("Message must be a JSON object")
  }
  return dict
}

private func writeNativeMessage(_ payload: [String: Any]) throws {
  let data = try JSONSerialization.data(withJSONObject: payload, options: [])
  var lengthLE = UInt32(data.count).littleEndian
  let header = Data(bytes: &lengthLE, count: 4)
  try FileHandle.standardOutput.write(contentsOf: header)
  try FileHandle.standardOutput.write(contentsOf: data)
}

private struct ProcessResult {
  let stdout: String
  let stderr: String
  let exitCode: Int32
}

private func runProcess(_ executable: String, _ args: [String], stdin: Data? = nil) throws -> ProcessResult {
  let proc = Process()
  proc.executableURL = URL(fileURLWithPath: executable)
  proc.arguments = args

  let outPipe = Pipe()
  let errPipe = Pipe()
  proc.standardOutput = outPipe
  proc.standardError = errPipe

  if let stdin {
    let inPipe = Pipe()
    proc.standardInput = inPipe
    try proc.run()
    inPipe.fileHandleForWriting.write(stdin)
    try inPipe.fileHandleForWriting.close()
  } else {
    try proc.run()
  }

  proc.waitUntilExit()

  let stdoutData = outPipe.fileHandleForReading.readDataToEndOfFile()
  let stderrData = errPipe.fileHandleForReading.readDataToEndOfFile()
  let stdout = String(data: stdoutData, encoding: .utf8) ?? ""
  let stderr = String(data: stderrData, encoding: .utf8) ?? ""

  return ProcessResult(stdout: stdout.trimmingCharacters(in: .whitespacesAndNewlines),
                       stderr: stderr.trimmingCharacters(in: .whitespacesAndNewlines),
                       exitCode: proc.terminationStatus)
}

private func runJXA(_ code: String) throws -> String {
  let res = try runProcess("/usr/bin/osascript", ["-l", "JavaScript", "-"], stdin: code.data(using: .utf8))
  if res.exitCode != 0 {
    throw NativeHostError.process(res.stderr.isEmpty ? "JXA failed (code \(res.exitCode))" : res.stderr)
  }
  return res.stdout
}

private func runAppleScriptFile(_ scriptPath: String, argv: [String]) throws -> String {
  let res = try runProcess("/usr/bin/osascript", [scriptPath] + argv)
  if res.exitCode != 0 {
    throw NativeHostError.process(res.stderr.isEmpty ? "osascript failed (code \(res.exitCode))" : res.stderr)
  }
  return res.stdout
}

private func escapeHtml(_ text: String) -> String {
  var out = text
  out = out.replacingOccurrences(of: "&", with: "&amp;")
  out = out.replacingOccurrences(of: "<", with: "&lt;")
  out = out.replacingOccurrences(of: ">", with: "&gt;")
  out = out.replacingOccurrences(of: "\"", with: "&quot;")
  out = out.replacingOccurrences(of: "'", with: "&#39;")
  return out
}

private func markdownToNotesHtml(_ markdown: String) -> String {
  let lines = markdown.components(separatedBy: .newlines)
  var out: [String] = []
  for line in lines {
    if line.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      out.append("<div><br></div>")
      continue
    }

    let trimmed = line.trimmingCharacters(in: .whitespaces)
    if trimmed.hasPrefix("<img ") && trimmed.hasSuffix(">") {
      out.append("<div>\(trimmed)</div>")
      continue
    }

    var prefixHtml = ""
    var rest = line
    if let range = line.range(of: #"^[ \t]+"#, options: .regularExpression) {
      let ws = String(line[range])
      rest = String(line[range.upperBound...])
      let expanded = ws.replacingOccurrences(of: "\t", with: "    ")
      prefixHtml = expanded.map { $0 == " " ? "&nbsp;" : String($0) }.joined()
    }
    out.append("<div>\(prefixHtml)\(escapeHtml(rest))</div>")
  }
  return out.joined(separator: "\n")
}

private func guessMime(from url: URL) -> String {
  switch url.pathExtension.lowercased() {
  case "png": return "image/png"
  case "jpg", "jpeg": return "image/jpeg"
  case "gif": return "image/gif"
  case "webp": return "image/webp"
  default: return "application/octet-stream"
  }
}

private func fileExtension(for mime: String, fallbackUrl: URL?) -> String {
  let m = mime.lowercased()
  switch m {
  case "image/png": return ".png"
  case "image/jpeg", "image/jpg": return ".jpg"
  case "image/gif": return ".gif"
  case "image/webp": return ".webp"
  default:
    if let ext = fallbackUrl?.pathExtension, !ext.isEmpty, ext.count <= 6 {
      return "." + ext
    }
    return ".bin"
  }
}

private func ensureCacheDir() throws -> URL {
  let dir = FileManager.default.temporaryDirectory.appendingPathComponent(cacheDirName, isDirectory: true)
  try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
  return dir
}

private func cleanupCacheDirIfPossible() {
  do {
    let dir = try ensureCacheDir()
    let now = Date()
    let contents = try FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: [.contentModificationDateKey], options: [.skipsHiddenFiles])
    for file in contents {
      let values = try file.resourceValues(forKeys: [.contentModificationDateKey])
      if let mtime = values.contentModificationDate, now.timeIntervalSince(mtime) > cacheTtlSeconds {
        try? FileManager.default.removeItem(at: file)
      }
    }
  } catch {
    // Ignore cache cleanup errors.
  }
}

private func downloadBytes(urlString: String, referer: String?) throws -> (data: Data, mime: String) {
  guard let url = URL(string: urlString) else { throw NativeHostError.download("Invalid URL: \(urlString)") }

  if url.isFileURL {
    let data = try Data(contentsOf: url)
    return (data, guessMime(from: url))
  }

  var req = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 60)
  req.setValue("Mozilla/5.0 (Macintosh; Intel Mac OS X) AppleWebKit/537.36 (KHTML, like Gecko) Chrome Safari", forHTTPHeaderField: "User-Agent")
  if let referer, !referer.isEmpty { req.setValue(referer, forHTTPHeaderField: "Referer") }

  let sem = DispatchSemaphore(value: 0)
  var resultData: Data?
  var resultMime: String?
  var resultErr: Error?

  URLSession.shared.dataTask(with: req) { data, response, error in
    defer { sem.signal() }
    if let error { resultErr = error; return }
    resultData = data
    if let http = response as? HTTPURLResponse {
      let ct = http.value(forHTTPHeaderField: "Content-Type") ?? ""
      let mime = ct.split(separator: ";", maxSplits: 1, omittingEmptySubsequences: true).first.map { String($0).trimmingCharacters(in: .whitespaces) }
      if let mime, !mime.isEmpty { resultMime = mime }
    }
  }.resume()

  if sem.wait(timeout: .now() + 60) == .timedOut {
    throw NativeHostError.timeout("Timeout downloading image: \(urlString)")
  }
  if let resultErr { throw NativeHostError.download(resultErr.localizedDescription) }
  guard let data = resultData else { throw NativeHostError.download("No data downloading image: \(urlString)") }
  let mime = resultMime ?? guessMime(from: url)
  return (data, mime)
}

private func cacheImageToFileUrl(urlString: String, referer: String?) throws -> URL {
  let (data, mime) = try downloadBytes(urlString: urlString, referer: referer)
  let dir = try ensureCacheDir()
  let ext = fileExtension(for: mime, fallbackUrl: URL(string: urlString))
  let filename = "img-\(UUID().uuidString)\(ext)"
  let path = dir.appendingPathComponent(filename)
  try data.write(to: path, options: .atomic)
  return path
}

private func replaceImageTokens(markdown: String, images: [[String: Any]], sourceUrl: String?) -> String {
  var out = markdown
  for img in images {
    guard let token = img["token"] as? String, let url = img["url"] as? String else { continue }
    let placeholder = "[[[IMG:\(token)]]]"
    do {
      let fileUrl = try cacheImageToFileUrl(urlString: url, referer: sourceUrl)
      out = out.replacingOccurrences(of: placeholder, with: "<img src=\"\(fileUrl.absoluteString)\">")
    } catch {
      out = out.replacingOccurrences(of: placeholder, with: url)
    }
  }
  return out
}

private func listFolders() throws -> [String: Any] {
  let code = #"""
    const Notes = Application('Notes');
    Notes.includeStandardAdditions = true;

    function collect(folder, prefix) {
      const name = folder.name();
      const path = prefix ? `${prefix}/${name}` : name;
      let out = [{ path }];
      const children = folder.folders();
      for (const child of children) out = out.concat(collect(child, path));
      return out;
    }

    const accounts = Notes.accounts().map(acc => {
      const folders = [];
      for (const f of acc.folders()) folders.push(...collect(f, ''));
      const filtered = folders.filter(x => x.path !== 'Recently Deleted');
      return { name: acc.name(), folders: filtered };
    });

    JSON.stringify({ ok: true, accounts });
  """#

  let raw = try runJXA(code)
  let obj = try JSONSerialization.jsonObject(with: Data(raw.utf8), options: [])
  guard let dict = obj as? [String: Any] else { throw NativeHostError.process("Invalid listFolders output") }
  return dict
}

private func createNote(msg: [String: Any]) throws -> [String: Any] {
  let title = (msg["title"] as? String) ?? "Untitled"
  let sourceUrl = (msg["sourceUrl"] as? String) ?? ""
  var markdown = (msg["markdown"] as? String) ?? ""
  markdown = markdown.trimmingCharacters(in: .whitespacesAndNewlines)

  let folder = (msg["folder"] as? [String: Any]) ?? [:]
  let accountName = (folder["accountName"] as? String) ?? ""
  let folderPath = (folder["folderPath"] as? String) ?? ""

  let images = (msg["images"] as? [[String: Any]]) ?? []
  if !images.isEmpty {
    markdown = replaceImageTokens(markdown: markdown, images: images, sourceUrl: sourceUrl)
  }

  let html = markdownToNotesHtml(markdown)
  let tmpDir = FileManager.default.temporaryDirectory
  let tmpUrl = tmpDir.appendingPathComponent("apple-notes-webclipper-\(UUID().uuidString).html")
  guard let htmlData = html.data(using: .utf8) else {
    throw NativeHostError.io("Failed to encode HTML as UTF-8")
  }
  try htmlData.write(to: tmpUrl, options: .atomic)

  defer { try? FileManager.default.removeItem(at: tmpUrl) }

  let exeUrl = URL(fileURLWithPath: CommandLine.arguments.first ?? ".")
  let scriptsDir = exeUrl.deletingLastPathComponent().appendingPathComponent("scripts")
  let scriptPath = scriptsDir.appendingPathComponent("create_note.applescript").path

  let noteId = try runAppleScriptFile(scriptPath, argv: [accountName, folderPath, title, tmpUrl.path])
  return ["ok": true, "noteId": noteId]
}

private func renderHtmlPreview(msg: [String: Any]) throws -> [String: Any] {
  let title = (msg["title"] as? String) ?? "Untitled"
  let sourceUrl = (msg["sourceUrl"] as? String) ?? ""
  var markdown = (msg["markdown"] as? String) ?? ""
  markdown = markdown.trimmingCharacters(in: .whitespacesAndNewlines)

  let images = (msg["images"] as? [[String: Any]]) ?? []
  if !images.isEmpty {
    markdown = replaceImageTokens(markdown: markdown, images: images, sourceUrl: sourceUrl)
  }

  let fragment = markdownToNotesHtml(markdown)
  let full = """
  <!doctype html>
  <html>
  <head>
    <meta charset="utf-8">
    <title>\(escapeHtml(title))</title>
  </head>
  <body>
  \(fragment)
  </body>
  </html>
  """

  let dir = try ensureCacheDir()
  let outUrl = dir.appendingPathComponent("preview-\(UUID().uuidString).html")
  guard let data = full.data(using: .utf8) else { throw NativeHostError.io("Failed to encode HTML as UTF-8") }
  try data.write(to: outUrl, options: .atomic)

  return ["ok": true, "htmlPath": outUrl.path, "htmlFileUrl": outUrl.absoluteString]
}

cleanupCacheDirIfPossible()

while true {
  do {
    guard let msg = try readNativeMessage() else { break }
    let action = (msg["action"] as? String) ?? ""

    switch action {
    case "ping":
      try writeNativeMessage(["ok": true])
    case "listFolders":
      try writeNativeMessage(try listFolders())
    case "createNote":
      try writeNativeMessage(try createNote(msg: msg))
    case "renderHtml":
      try writeNativeMessage(try renderHtmlPreview(msg: msg))
    default:
      try writeNativeMessage(["ok": false, "error": "Unknown action: \(action)"])
    }
  } catch {
    let errMsg = (error as? NativeHostError)?.description ?? String(describing: error)
    try? writeNativeMessage(["ok": false, "error": errMsg])
  }
}
